### PR TITLE
Implemented Coca architecture

### PR DESCRIPTION
**Pull request by VarunS1997**
# What does this PR do?

Implements the work done in the "CoCa": Contrastive Captioners are Image-Text Foundation Models" (https://arxiv.org/pdf/2205.01917.pdf). 

This PR requires: 
 [x] Model Implementation
 [ ] Testing

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/keras-team/keras-cv/blob/master/.github/CONTRIBUTING.md),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue? Please add a link
      to it if that's the case.
- [ ] Did you write any new necessary tests?
- [ ] If this adds a new model, can you run a few training steps on TPU in Colab to ensure that no XLA incompatible OP are used?

### CODE REVIEW

File: keras_cv/models/feature_extractor/CoCa/coca_model.py [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+        self.con_loss_weight = con_loss_weight
+
+        self.cap_queries = cap_queries
+        self.cap_heads = cap_heads
+        self.cap_loss_weight = cap_loss_weight
+
+    def build(self, input_shape):
+        super().build(input_shape)
+
+        self.image_patching = PatchingAndEmbedding(self.encoder_width, self.img_patch_size)
```
Comment (divyashreepathihalli) [inline keras_cv/models/feature_extractor/CoCa/coca_model.py:74]: lets move all the layer definitions to `__init__()`

File: keras_cv/models/feature_extractor/CoCa/coca_model.py [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+        self.image_encoder = Sequential([
+            CVTransformerEncoder(self.img_query_dim, self.encoder_heads, self.encoder_intermediate_dim)
+            for _ in range(self.encoder_depth)
```
Comment (divyashreepathihalli) [inline keras_cv/models/feature_extractor/CoCa/coca_model.py:77]: And build all the layers here like `layer.build(<expected_input_shape>)`

**Review by divyashreepathihalli — CHANGES_REQUESTED** [review]
This is great @VarunS1997 !! left a few comments. Also, please add a colab to verify the output.

File: keras_cv/models/feature_extractor/CoCa/coca_model.py [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+                 encoder_depth=40,
+                 encoder_heads=16,
+                 encoder_intermediate_dim=6144,
+                 encoder_width=1408,
+                 unimodal_decoder_depth=18,
+                 multimodal_decoder_depth=18,
+                 decoder_intermediate_dim=5632,
+                 unimodal_decoder_heads=16,
+                 multimodal_decoder_heads=16,
+                 con_queries=1,
```
Comment (divyashreepathihalli) [inline keras_cv/models/feature_extractor/CoCa/coca_model.py:39]: lets expand `con` -> `contrastive` here and everywhere.

**Review by divyashreepathihalli — CHANGES_REQUESTED** [review]

**Comment by divyashreepathihalli** [comment]
One additional overhead work is needed.

please add keras_cv/models/feature_extractor/coca  to this file
https://github.com/keras-team/keras-cv/blob/master/.kokoro/github/ubuntu/gpu/build.sh
to line 72 and 86

PS: we will fix this overhead soon, but in the mean time this is what we need to do to make sure the large GPU tests run.

File: keras_cv/layers/attention_pooling.py [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+
+class AttentionPooling(layers.Layer):
+    """Implements the Pooled Attention Layer used in "coca": Contrastive Captioners are Image-Text Foundation Models"
+    (https://arxiv.org/pdf/2205.01917.pdf), consisting of a Multiheaded Attention followed by Layer Normalization.
+
+    Args:
+        proj_dim: The dimensions of the attention heads
+        num_heads: The number of attention heads in the multi-headed attention layer
+    """
+    def __init__(self,
```
Comment (mattdangerw) [inline keras_cv/layers/attention_pooling.py:12]: we should make sure to format with `black`, this looks like another formatting

File: keras_cv/models/feature_extractor/coca/coca_model.py [hunk 5]
```diff
@@ -0,0 +1,244 @@
...
+        ])
+
+        self.contrastive_attn_pooling = AttentionPooling(self.encoder_width, self.contrastive_attn_heads)
+        self.captioning_attn_pooling = AttentionPooling(self.encoder_width, self.captioning_attn_heads)
+
+        # These are learnable weights defined in build as per Keras recommendations
+        self.cls_token = None
+        self.contrastive_query = None
+        self.captioning_query = None
+    """ Contrastive Captioner foundational model implementation.
```
Comment (mattdangerw) [inline keras_cv/models/feature_extractor/coca/coca_model.py:91]: docstring should go above `__init__` after the class line

File: keras_cv/models/feature_extractor/coca/coca_model.py [hunk 6]
```diff
@@ -0,0 +1,244 @@
...
+
+        # This is only needed for loss calculations
+        # contrastive_feature = self.con_attn_pooling(self.contrastive_query, img_encoding)
+
+        # [batch_size, encoder_width, captioning_query_length]
+        captioning_feature = self.captioning_attn_pooling(self.captioning_query, img_encoding)
+
+        # [batch_size, sequence_length+1, text_dim]
+        text_tokens = np.concatenate(texts, self.cls_token)
+        mask = np.concatenate((np.ones_like(texts), np.zeros_like(self.cls_token)))
```
Comment (mattdangerw) [inline keras_cv/models/feature_extractor/coca/coca_model.py:209]: I can't tell if this mask is the right shape or not. Usually you want something `(batch_size, seq_length, seq_length)` (or seq lenth + 1 if that is the effective sequence length. What is `text_dim` here?

**Review by mattdangerw — COMMENTED** [review]
Thanks! Left a few comments.

File: keras_cv/models/feature_extractor/coca/coca_model.py (hunk 6 above)
Comment (VarunS1997) [reply keras_cv/models/feature_extractor/coca/coca_model.py:209]: text_dim is the dimensionality of the text embeddings

File: keras_cv/models/feature_extractor/coca/coca_model.py [hunk 7]
```diff
@@ -0,0 +1,244 @@
...
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import numpy as np
+from keras import Sequential
+from keras_cv.api_export import keras_cv_export
+from keras_nlp.layers import RotaryEmbedding, TransformerDecoder
```
[the remark this answers is not in the document]
Comment (divyashreepathihalli) [reply keras_cv/models/feature_extractor/coca/coca_model.py:17]: I think as we support more multi modal models we should depend on KerasNLP. If the tf-text install issue is resolved, we should add it.

File: keras_cv/models/feature_extractor/coca/coca_model.py [hunk 8]
```diff
@@ -0,0 +1,244 @@
...
+        img_encoding = self.image_encoder(img_encoding)  # [batch_size, img_patches_len+1, encoder_width]
+
+        # This is only needed for loss calculations
+        # contrastive_feature = self.con_attn_pooling(self.contrastive_query, img_encoding)
+
+        # [batch_size, encoder_width, captioning_query_length]
+        captioning_feature = self.captioning_attn_pooling(self.captioning_query, img_encoding)
+
+        # [batch_size, sequence_length+1, text_dim]
+        text_tokens = np.concatenate(texts, self.cls_token)
```
[the remark this answers is not in the document]
Comment (VarunS1997) [reply keras_cv/models/feature_extractor/coca/coca_model.py:208]: Yeah, I considered that but I think because the token doesn't actually leave the model's intermediate representation, it's okay to leave here.

Otherwise, it would create a weird API where the input is of length N, but the output is length N-1.

File: keras_cv/layers/attention_pooling.py [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+            self.num_heads, self.head_dim
+        )
+
+        self.layer_norm = layers.LayerNormalization()
+
+    def build(self, input_shape):
+        super().build(input_shape)
+
+        self.multi_head_attn.build(input_shape)
+        self.layer_norm.build(input_shape)
```
Comment (divyashreepathihalli) [inline keras_cv/layers/attention_pooling.py:29]: is the input shape for layer_norm correct?

File: keras_cv/models/feature_extractor/coca/coca_model.py [hunk 10]
```diff
@@ -0,0 +1,311 @@
...
+
+        self.captioning_query_length = captioning_query_length
+        self.captioning_attn_heads = captioning_attn_heads
+        self.captioning_loss_weight = captioning_loss_weight
+
+        # Layer Definitions
+        self.image_patching = PatchingAndEmbedding(
+            self.encoder_width, self.img_patch_size
+        )
+        self.image_encoder = Sequential(
```
Comment (divyashreepathihalli) [inline keras_cv/models/feature_extractor/coca/coca_model.py:119]: Sequential might not work, the model will not build properly.

File: keras_cv/models/feature_extractor/coca/coca_model.py [hunk 11]
```diff
@@ -0,0 +1,311 @@
...
+                    self.encoder_width,
+                    self.encoder_heads,
+                    self.encoder_intermediate_dim,
+                )
+                for _ in range(self.encoder_depth)
+            ]
+        )
+
+        self.text_embedding = RotaryEmbedding()
+        self.unimodal_text_decoder = Sequential(
```
Comment (divyashreepathihalli) [inline keras_cv/models/feature_extractor/coca/coca_model.py:131]: Again, sequential might not work well. Please double check.

File: keras_cv/models/feature_extractor/coca/coca_model.py [hunk 12]
```diff
@@ -0,0 +1,311 @@
...
+            [
+                TransformerDecoder(
+                    self.decoder_intermediate_dim, self.unimodal_decoder_heads
+                )
+                for _ in range(self.unimodal_decoder_depth)
+            ]
+        )
+        self.multimodal_text_decoder = Sequential(
```
Comment (divyashreepathihalli) [inline keras_cv/models/feature_extractor/coca/coca_model.py:139]: same comment about Sequential

File: keras_cv/models/feature_extractor/coca/coca_model.py [hunk 13]
```diff
@@ -0,0 +1,311 @@
...
+            )
+
+        # Build Layers
+        self.image_patching.build(images_shape)
+
+        # Add 1 for CLs token appended by patching
+        num_patches = (images_shape[1] // self.img_patch_size) * (
+            images_shape[2] // self.img_patch_size
+        ) + 1
+        self.image_encoder.build((batch_size, self.encoder_width, num_patches))
```
Comment (divyashreepathihalli) [inline keras_cv/models/feature_extractor/coca/coca_model.py:198]: you could keep batch_size as `None`
example 
`self.image_encoder.build((None, self.encoder_width, num_patches))`

**Review by divyashreepathihalli — CHANGES_REQUESTED** [review]
Thanks! I left a few comments
Also, please do add a colab demo to show that the model is working as expected.

File: keras_cv/layers/attention_pooling.py [hunk 14]
```diff
@@ -0,0 +1,33 @@
+from keras import layers
+
+
+class AttentionPooling(layers.Layer):
+    """Implements the Pooled Attention Layer used in "coca": Contrastive Captioners are Image-Text Foundation Models"
+    (https://arxiv.org/pdf/2205.01917.pdf), consisting of a Multiheaded Attention followed by Layer Normalization.
```
Comment (innat-asj) [inline keras_cv/layers/attention_pooling.py:6]: You probably need to rename this pooling layer for COCO. There are many ways to implement attention pooling layer and so this could be confusing later. Or, you can define a separate layers dedicated for coco models.

layers/coco_layers.py

define AttentionPooling in the coco_layers.py

File: keras_cv/models/feature_extractor/coca/coca_model.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (mattdangerw) [reply keras_cv/models/feature_extractor/coca/coca_model.py]: Sgtm! Though if we switch to a hard dependency here, we should probably add `keras-nlp` as a dependency in `setup.py` (which comes with a transitive dependency on `tensorflow-text` and `tensorflow` just fyi).

[the remark this answers is not in the document]
Comment (VarunS1997) [reply keras_cv/models/feature_extractor/coca/coca_model.py]: Do we want to include that in this PR? There's already some imports of Keras-NLP in other places of Keras CV.

If we make it a separate PR, it'll make it easier to rollback if we need to. Considering it's a new dependency, might be worth separating.

File: keras_cv/models/feature_extractor/coca/coca_model.py (hunk 13 above)
Comment (VarunS1997) [reply keras_cv/models/feature_extractor/coca/coca_model.py:198]: Just for my understanding, is there a specific reason to do that over setting the batch_size?

File: keras_cv/layers/attention_pooling.py (hunk 9 above)
Comment (VarunS1997) [reply keras_cv/layers/attention_pooling.py]: Fixed it, let me know if it still doesn't look right!

File: keras_cv/models/feature_extractor/coca/coca_model.py [hunk 15]
```diff
@@ -0,0 +1,279 @@
...
+    This model implements the "Contrastive Captioners are image-Text Foundational Models" by Yu, et al.
+    (https://arxiv.org/pdf/2205.01917.pdf). In short, the coca model combines the ideas of Contrastive techniques
+    such as CLIP, with Generative Captioning approaches such as SimVLM.
+
+    The architecture of clip can be described as an Image Visual Transformer Encoder in parallel to self-attention-only
+    Text Transformer Decoder, the outputs of both of which are passed into a multimodal Transformer Decoder. The
+    contrastive loss from the ViT and the uni-modal Text Decoder is combined with a captioning loss from the multi-modal
+    Decoder in order to produce the combined total loss.
+
+    Basic Usage:
```
Comment (sachinprasadhs) [inline keras_cv/models/feature_extractor/coca/coca_model.py:40]: This has to be changed to `Example:` since we follow only `Example` or `Examples:` as a standard format.

**Comment by divyashreepathihalli** [comment]
@VarunS1997 will you be completing this one?
